src/combinatorics/quiver/max_flow_min_cut.lean

```diff
@@ -0,0 +1,1507 @@
+begin
+  intro eq,
+  intros cut eq_net,
+  rw eq,
+  have h: afn.network = cut.network := by {rw same_net, exact eq_net,},
+  exact weak_duality (afn) (cut) (h),
+end
+
+/-!
+  We now define our residual network. Our goal is to prove no_augm_path.
+-/
+
+noncomputable
+def mk_rsf {V : Type*}  [inst' : fintype V] -- stays for residual flow
+  (afn : active_flow_network V)
+  (u v : V) : ℝ :=
+  if  afn.network.is_edge u v
+  then afn.network.c u v - afn.f u v
+  else if afn.network.is_edge v u
+        then afn.f v u
+        else 0
+
+structure residual_network  (V : Type*)  [inst' : fintype V] :=
+  (afn : active_flow_network V)
+  (f' : V -> V -> ℝ)
+  (f_def : f' = mk_rsf afn)
+  (is_edge : V -> V -> Prop)
+  (is_edge_def : is_edge = λ u v, f' u v > 0)
+
+noncomputable
+def mk_rsn {V : Type*} [fintype V] -- stays for residual network
+  (afn : active_flow_network V) : residual_network V :=
+  ⟨afn, mk_rsf afn, rfl, λ u v, mk_rsf afn u v > 0 , rfl⟩
+
+universe u
+
+/-
+  We define a recursive structure returning the path between two vertices,
+  if such a path exists, given the edges in the graph.
+-/
+inductive path {V : Type u } (is_edge : V -> V -> Prop) (a : V) : V → Type (u + 1)
+| nil  : path a
+| cons : Π {b c : V}, path b → (is_edge b c) → path c
+
+def no_augmenting_path {V : Type*} [inst' : fintype V]
+  (rsn : residual_network V) : Prop :=
+  ∀ t : V, ∀ p : path rsn.is_edge rsn.afn.network.source t, ¬ (t = rsn.afn.network.sink)
+
+/-
+  Given two vertices $u and $v, returns true if and only if (u,v) is an edge in a given path.
+-/
+def path.in {V : Type u}
+  {is_edge : V -> V -> Prop}
+  (u v : V)
+  {s : V} : ∀ {t : V}, path is_edge s t -> Prop
+  | t (@path.nil  _ is_edge' a)  := u = v
+  | t (@path.cons _ _ _ t' _ p _)  := (u = t' ∧ v = t) ∨ (@path.in t' p)
+
+lemma residual_capacity_non_neg {V : Type*} [inst' : fintype V]
+  (rsn : residual_network V) : ∀ u v : V,  0 ≤ rsn.f' u v :=
+begin
+  intros u v,
+  cases rsn,
+  simp only,
+  rw rsn_f_def,
+  unfold mk_rsf,
+  have foo := classical.em (rsn_afn.network.to_capacity.to_digraph.is_edge u v),
+  cases foo,
+  { simp only [foo, if_true, sub_nonneg, rsn_afn.no_overflow], },
+  { simp only [foo, if_false], clear foo,
+    have bar := classical.em (rsn_afn.network.to_capacity.to_digraph.is_edge v u),
+    cases bar,
+    { have h := rsn_afn.non_neg_flow v u,
+      simp only [bar, h, if_true],
+      linarith, },
+    { simp only [bar, if_false] }, },
+end
+
+lemma positive_residual_flow {V : Type*} [inst' : fintype V]
+  (rsn : residual_network V) (u v : V) : rsn.is_edge u v → rsn.f' u v > 0 :=
+begin
+  intro edge,
+  rw rsn.is_edge_def at edge,
+  exact edge,
+end
+
+-- /-!
+--   Here is our second big lemma, if the active flow is maximum,
+--   then no augmenting path exists in the residual network.
+-- -/
+
+-- lemma no_augm_path {V : Type*} [inst' : fintype V]
+--   (rsn : residual_network V) : (is_max_flow rsn.afn) -> no_augmenting_path rsn :=
+-- begin
```

Comment:
If this is intended to be part of a later PR, you should just delete it from this one.